File: server/observability/metricsStore.mjs
const MAX_LATENCY_SAMPLES = 2000;
const MAX_ROUTE_ENTRIES = 200;

const state = {
  startedAt: new Date().toISOString(),
  totalRequests: 0,
  totalErrors: 0,
  latencies: [],
  statusClasses: {
    '2xx': 0,
    '3xx': 0,
    '4xx': 0,
    '5xx': 0,
    other: 0,
  },
  routes: new Map(),
  errorCodes: new Map(),
  fallbackByModule: new Map(),
};

function classifyStatus(status) {
  if (status >= 200 && status < 300) return '2xx';
  if (status >= 300 && status < 400) return '3xx';
  if (status >= 400 && status < 500) return '4xx';
  if (status >= 500 && status < 600) return '5xx';
  return 'other';
}

function routeKey(method, path) {
  return `${String(method || 'GET').toUpperCase()} ${String(path || '/')}`;
}

function ensureRouteEntry(key) {
  if (!state.routes.has(key)) {
    if (state.routes.size >= MAX_ROUTE_ENTRIES) {
      return null;
    }
    state.routes.set(key, {
      requests: 0,
      errors: 0,
      latencySumMs: 0,
      maxLatencyMs: 0,
    });
  }
  return state.routes.get(key);
}

function pushLatencySample(value) {
  state.latencies.push(value);
  if (state.latencies.length > MAX_LATENCY_SAMPLES) {
    state.latencies.shift();
  }
}

function percentile(values, p) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return sorted[idx];
}

export function resetObservabilityMetrics() {
  state.startedAt = new Date().toISOString();
  state.totalRequests = 0;
  state.totalErrors = 0;
  state.latencies = [];
  state.statusClasses = {
    '2xx': 0,
    '3xx': 0,
    '4xx': 0,
    '5xx': 0,
    other: 0,
  };
  state.routes.clear();
  state.errorCodes.clear();
  state.fallbackByModule.clear();
}

export function recordApiRequestMetric({ method, path, status, latencyMs }) {
  const safeStatus = Number(status || 0) || 0;
  const safeLatency = Number(latencyMs || 0) || 0;

  state.totalRequests += 1;
  state.statusClasses[classifyStatus(safeStatus)] += 1;
  pushLatencySample(safeLatency);

  if (safeStatus >= 400) {
    state.totalErrors += 1;
  }

  const key = routeKey(method, path);
  const entry = ensureRouteEntry(key);
  if (!entry) {
    return;
  }

  entry.requests += 1;
  entry.latencySumMs += safeLatency;
  entry.maxLatencyMs = Math.max(entry.maxLatencyMs, safeLatency);
  if (safeStatus >= 400) {
    entry.errors += 1;
  }
}

export function recordApiErrorCodeMetric(code) {
  const key = String(code || 'UNKNOWN_ERROR');
  state.errorCodes.set(key, (state.errorCodes.get(key) || 0) + 1);
}

export function recordApiFallbackMetric(moduleName, reason = 'unknown') {
  const moduleKey = String(moduleName || 'unknown').trim().toLowerCase() || 'unknown';
  const reasonKey = String(reason || 'unknown').trim().toLowerCase() || 'unknown';

  if (!state.fallbackByModule.has(moduleKey)) {
    state.fallbackByModule.set(moduleKey, {
      count: 0,
      reasons: new Map(),
    });
  }

  const moduleEntry = state.fallbackByModule.get(moduleKey);
  moduleEntry.count += 1;
  moduleEntry.reasons.set(reasonKey, (moduleEntry.reasons.get(reasonKey) || 0) + 1);
}

export function getObservabilityMetricsSnapshot({ routeLimit = 10, codeLimit = 10 } = {}) {
  const avgLatency = state.latencies.length > 0
    ? Number((state.latencies.reduce((sum, value) => sum + value, 0) / state.latencies.length).toFixed(2))
    : 0;
  const p95Latency = Number(percentile(state.latencies, 95).toFixed(2));
  const maxLatency = state.latencies.length > 0 ? Math.max(...state.latencies) : 0;
  const errorRate = state.totalRequests > 0
    ? Number(((state.totalErrors / state.totalRequests) * 100).toFixed(2))
    : 0;

  const topRoutes = [...state.routes.entries()]
    .map(([route, entry]) => ({
      route,
      requests: entry.requests,
      errors: entry.errors,
      avgLatencyMs: entry.requests > 0 ? Number((entry.latencySumMs / entry.requests).toFixed(2)) : 0,
      maxLatencyMs: entry.maxLatencyMs,
    }))
    .sort((a, b) => b.requests - a.requests)
    .slice(0, Math.max(1, Number(routeLimit || 10)));

  const errorCodes = [...state.errorCodes.entries()]
    .map(([code, count]) => ({ code, count }))
    .sort((a, b) => b.count - a.count)
    .slice(0, Math.max(1, Number(codeLimit || 10)));

  const fallbackModules = [...state.fallbackByModule.entries()]
    .map(([module, entry]) => ({
      module,
      count: entry.count,
      reasons: [...entry.reasons.entries()]
        .map(([reason, count]) => ({ reason, count }))
        .sort((a, b) => b.count - a.count),
    }))
    .sort((a, b) => b.count - a.count);
  const totalFallbacks = fallbackModules.reduce((acc, item) => acc + item.count, 0);

  return {
    generatedAt: new Date().toISOString(),
    startedAt: state.startedAt,
    counters: {
      totalRequests: state.totalRequests,
      totalErrors: state.totalErrors,
      errorRatePct: errorRate,
    },
    latency: {
      avgMs: avgLatency,
      p95Ms: p95Latency,
      maxMs: maxLatency,
      samples: state.latencies.length,
    },
    statusClasses: { ...state.statusClasses },
    topRoutes,
    errorCodes,
    fallbacks: {
      total: totalFallbacks,
      modules: fallbackModules,
    },
  };
}

File: backend/app/observability/metrics_store.py
from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime, timezone
from statistics import mean
from typing import Any

MAX_LATENCY_SAMPLES = 2000
MAX_ROUTE_ENTRIES = 200


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass
class RouteMetric:
    requests: int = 0
    errors: int = 0
    latency_sum_ms: float = 0
    max_latency_ms: float = 0


@dataclass
class MetricsState:
    started_at: str = field(default_factory=_now_iso)
    total_requests: int = 0
    total_errors: int = 0
    latencies: list[float] = field(default_factory=list)
    status_classes: Counter = field(default_factory=lambda: Counter({"2xx": 0, "3xx": 0, "4xx": 0, "5xx": 0, "other": 0}))
    routes: dict[str, RouteMetric] = field(default_factory=dict)
    error_codes: Counter = field(default_factory=Counter)
    fallback_by_module: dict[str, dict[str, Any]] = field(default_factory=dict)


state = MetricsState()


def reset_observability_metrics() -> None:
    global state
    state = MetricsState()


def _classify_status(status: int) -> str:
    if 200 <= status < 300:
        return "2xx"
    if 300 <= status < 400:
        return "3xx"
    if 400 <= status < 500:
        return "4xx"
    if 500 <= status < 600:
        return "5xx"
    return "other"


def record_api_request_metric(method: str, path: str, status: int, latency_ms: float) -> None:
    state.total_requests += 1
    state.status_classes[_classify_status(status)] += 1
    if status >= 400:
        state.total_errors += 1

    state.latencies.append(float(latency_ms))
    if len(state.latencies) > MAX_LATENCY_SAMPLES:
        state.latencies.pop(0)

    key = f"{(method or 'GET').upper()} {path or '/'}"
    if key not in state.routes and len(state.routes) >= MAX_ROUTE_ENTRIES:
        return

    route = state.routes.setdefault(key, RouteMetric())
    route.requests += 1
    route.latency_sum_ms += float(latency_ms)
    route.max_latency_ms = max(route.max_latency_ms, float(latency_ms))
    if status >= 400:
        route.errors += 1


def record_api_error_code_metric(code: str) -> None:
    state.error_codes[code or "UNKNOWN_ERROR"] += 1


def record_api_fallback_metric(module_name: str, reason: str = "unknown") -> None:
    module_key = (module_name or "unknown").strip().lower() or "unknown"
    reason_key = (reason or "unknown").strip().lower() or "unknown"
    item = state.fallback_by_module.setdefault(module_key, {"count": 0, "reasons": Counter()})
    item["count"] += 1
    item["reasons"][reason_key] += 1


def _pct(values: list[float], p: int) -> float:
    if not values:
        return 0
    sorted_values = sorted(values)
    idx = min(len(sorted_values) - 1, int((p / 100) * len(sorted_values)))
    return float(sorted_values[idx])


def get_observability_metrics_snapshot(route_limit: int = 10, code_limit: int = 10) -> dict[str, Any]:
    avg = round(mean(state.latencies), 2) if state.latencies else 0
    p95 = round(_pct(state.latencies, 95), 2)
    max_latency = max(state.latencies) if state.latencies else 0
    error_rate = round((state.total_errors / state.total_requests) * 100, 2) if state.total_requests else 0

    top_routes = [
        {
            "route": key,
            "requests": route.requests,
            "errors": route.errors,
            "avgLatencyMs": round(route.latency_sum_ms / route.requests, 2) if route.requests else 0,
            "maxLatencyMs": route.max_latency_ms,
        }
        for key, route in state.routes.items()
    ]
    top_routes.sort(key=lambda x: x["requests"], reverse=True)

    error_codes = [{"code": code, "count": count} for code, count in state.error_codes.items()]
    error_codes.sort(key=lambda x: x["count"], reverse=True)

    modules = []
    for module, data in state.fallback_by_module.items():
        reasons = [{"reason": reason, "count": count} for reason, count in data["reasons"].items()]
        reasons.sort(key=lambda x: x["count"], reverse=True)
        modules.append({"module": module, "count": data["count"], "reasons": reasons})
    modules.sort(key=lambda x: x["count"], reverse=True)

    return {
        "generatedAt": _now_iso(),
        "startedAt": state.started_at,
        "counters": {
            "totalRequests": state.total_requests,
            "totalErrors": state.total_errors,
            "errorRatePct": error_rate,
        },
        "latency": {
            "avgMs": avg,
            "p95Ms": p95,
            "maxMs": max_latency,
            "samples": len(state.latencies),
        },
        "statusClasses": dict(state.status_classes),
        "topRoutes": top_routes[: max(1, route_limit)],
        "errorCodes": error_codes[: max(1, code_limit)],
        "fallbacks": {"total": sum(m["count"] for m in modules), "modules": modules},
    }
